pkg/clusterapi/yaml/controlplane.go

```diff
@@ -0,0 +1,149 @@
+package yaml
+
+import (
+	"github.com/go-logr/logr"
+	etcdv1 "github.com/mrajashree/etcdadm-controller/api/v1beta1"
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+
+	"github.com/aws/eks-anywhere/pkg/clusterapi"
+	"github.com/aws/eks-anywhere/pkg/yamlutil"
+)
+
+// NewControlPlaneParserAndBuilder builds a Parser and a Builder for a particular provider ControlPlane
+// It registers the basic shared mappings plus another two for the provider cluster and machine template
+// For ControlPlane that need to include more objects, wrap around the provider builder and implement BuildFromParsed
+// Any extra mappings will need to be registered manually in the Parser
+func NewControlPlaneParserAndBuilder[C, M clusterapi.Object](logger logr.Logger, clusterMapping yamlutil.Mapping[C], machineConfigMapping yamlutil.Mapping[M]) (*yamlutil.Parser, *ControlPlaneBuilder[C, M], error) {
+	parser := yamlutil.NewParser(logger)
+	if err := RegisterControlPlaneMappings(parser); err != nil {
+		return nil, nil, errors.Wrap(err, "building capi control plane parser")
+	}
+
+	err := parser.RegisterMappings(
+		clusterMapping.ToAPIObjectMapping(),
+		machineConfigMapping.ToAPIObjectMapping(),
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "registering provider control plane mappings")
+	}
+
+	return parser, NewControlPlaneBuilder[C, M](), nil
```

Comment:
why need to return parser? can't the provider initializes its own parser, registers mappings and update the `ControlPlane`? might be me not going through the POC draft completely

Reply:
Yeah, all of that is possible, that's why the mappings and process methods are public, they are building blocks. But this method is like a helper, it puts everything together following the most common case. Providers can choose to use it or implement their own.